### PR TITLE
[Feature] Separated circuit and observable parameters for AD

### LINKED ISSUE #188: [Feature, Refactoring, Differentiability] Split embedded params of circuit and observable
To support parametric observables for adjoint and GPSR, we can not natively compute gradients w.r.t. observable parameters. Observable parameters are purely classical and they should be treated separately using automatic differentiation. 

### Proposed solution

If we maintain two separate dictionaries for both, we can use AD on the observable. Notice that this refactoring will imply changes into the inner workings of `qadence` parametrization and will likely be quite extensive.

### PR DESCRIPTION
**Pull request by chMoussa**
Closes #188.

So far only for the PyQtorch backend.

### CODE REVIEW

File: qadence/types.py [hunk 1]
```diff
@@ -431,6 +431,7 @@ class ReadOutOptimization(StrEnum):
 
 
 ParamDictType = dict[str, ArrayLike]
+SeparatedParamDictType = dict[str, dict[str, ArrayLike]]
```
Comment (RolandMacDoland) [inline qadence/types.py:434]: Hm, not sure you really need another custom type. Can't you get away with just:

```python
ParamDictType = dict[str, ArrayLike | dict[str, ArrayLike]]
```
?

File: qadence/backend.py [hunk 2]
```diff
@@ -25,7 +25,15 @@
...
 from qadence.parameters import stringify
-from qadence.types import ArrayLike, BackendName, DiffMode, Endianness, Engine, ParamDictType
+from qadence.types import (
+    ArrayLike,
+    BackendName,
+    DiffMode,
+    Endianness,
+    Engine,
+    ParamDictType,
+    SeparatedParamDictType,
```
Comment (RolandMacDoland) [inline qadence/backend.py:35]: Not sure you really need yet another type. See below.

File: qadence/backend.py [hunk 3]
```diff
@@ -228,9 +236,15 @@ def check_observable(obs_obj: Any) -> AbstractBlock:
...
             def embedding_fn_dict(a: dict, b: dict) -> dict:
-                embedding_dict = circ_embedding_fn(a, b)
-                for o in obs_embedding_fn_list:
-                    embedding_dict.update(o(a, b))
+                if "circuit" in b:
+                    embedding_dict = {"circuit": circ_embedding_fn(a, b), "observables": dict()}
+                    if "observables" in b:
+                        for o in obs_embedding_fn_list:
+                            embedding_dict["observables"].update(o(a, b))
+                else:
```
Comment (RolandMacDoland) [inline qadence/backend.py:244]: Do I understand this clause as being "if the passed dict of parameters is not a dict of dicts" ? In that case, I'd make it more explicit because what if a user wants to differentiate over hamiltonian parameters only ?

File: qadence/backend.py [hunk 4]
```diff
@@ -228,9 +236,15 @@ def check_observable(obs_obj: Any) -> AbstractBlock:
...
+                    embedding_dict = circ_embedding_fn(a, b)
+                    for o in obs_embedding_fn_list:
```
Comment (RolandMacDoland) [inline qadence/backend.py:246]:
```suggestion
                    for obs_embedding_fn in obs_embedding_fns:
```

I'd use the plural as opposed to a `_list` suffix.

File: qadence/backends/horqrux/backend.py [hunk 5]
```diff
@@ -135,18 +135,29 @@ def expectation(
             A jax.Array of shape (batch_size, n_observables)
         """
         observable = observable if isinstance(observable, list) else [observable]
-        batch_size = max([arr.size for arr in param_values.values()])
+        if "observables" in param_values or "circuit" in param_values:
+            raise NotImplementedError("The HORQRUX backend does not support separated parameters.")
```
Comment (RolandMacDoland) [inline qadence/backends/horqrux/backend.py:139]:
```suggestion
            raise NotImplementedError("The Horqrux backend does not support separated parameters.")
```

File: qadence/backends/horqrux/backend.py [hunk 6]
```diff
@@ -135,18 +135,29 @@ def expectation(
...
+        def _expectation(params: ParamDictType | SeparatedParamDictType) -> ArrayLike:
+            param_circuits = params["circuit"] if "circuit" in params else params
+            param_observables = params["observables"] if "observables" in params else params
             out_state = self.run(
-                circuit, params, state, endianness, horqify_state=True, unhorqify_state=False
+                circuit,
+                param_circuits,
+                state,
+                endianness,
+                horqify_state=True,
```
Comment (RolandMacDoland) [inline qadence/backends/horqrux/backend.py:153]: Fair enough.

File: qadence/backends/horqrux/backend.py [hunk 7]
```diff
@@ -135,18 +135,29 @@ def expectation(
...
+                unhorqify_state=False,
             )
-            return jnp.array([o.native(out_state, params) for o in observable])
+            return jnp.array([o.native(out_state, param_observables) for o in observable])
```
Comment (RolandMacDoland) [inline qadence/backends/horqrux/backend.py:156]:
```suggestion
            return jnp.array([observable.native(out_state, param_observables) for observable in observables])
```

I guess we don't have to be stingy on names ;)

File: qadence/backends/pulser/backend.py [hunk 8]
```diff
@@ -322,38 +322,45 @@ def expectation(
         self,
         circuit: ConvertedCircuit,
         observable: list[ConvertedObservable] | ConvertedObservable,
-        param_values: dict[str, Tensor] = {},
+        param_values: dict[str, Tensor] | dict[str, dict[str, Tensor]] = {},
```
Comment (RolandMacDoland) [inline qadence/backends/pulser/backend.py:325]: No use of a `ParamDictType` ?

File: qadence/backends/pyqtorch/backend.py [hunk 9]
```diff
@@ -200,15 +205,15 @@ def _batched_expectation(
...
+            [obs.native.expectation(state, param_observables).reshape(-1, 1) for obs in observable]
         )
         return _expectation
 
     def _looped_expectation(
         self,
         circuit: ConvertedCircuit,
         observable: list[ConvertedObservable] | ConvertedObservable,
-        param_values: dict[str, Tensor] = {},
+        param_values: dict[str, Tensor] | dict[str, dict[str, Tensor]] = {},
```
Comment (RolandMacDoland) [inline qadence/backends/pyqtorch/backend.py:216]: `ParamDictType` ?

File: qadence/backends/pyqtorch/backend.py [hunk 10]
```diff
@@ -242,7 +256,7 @@ def expectation(
...
```
Comment (RolandMacDoland) [inline qadence/backends/pyqtorch/backend.py:259]: `ParamDictType` ?

**Review by RolandMacDoland — COMMENTED** [review]
Thanks @chMoussa. Generally speaking, I don't think the introduction of yet another custom type makes it simpler. Comments from my side.

File: qadence/types.py (hunk 1 above)
Comment (chMoussa) [reply qadence/types.py]: Yes I can!

File: qadence/types.py [hunk 11]
```diff
@@ -430,7 +430,7 @@ class ReadOutOptimization(StrEnum):
     CONSTRAINED = "constrained"
 
 
-ParamDictType = dict[str, ArrayLike]
+ParamDictType = dict[str, Union[ArrayLike, dict[str, ArrayLike]]]
```
Comment (RolandMacDoland) [inline qadence/types.py:433]: The more I see this, the more I think we should extend `ParamDictType` to a dataclass that will enforce the user passing circuit parameters but would default observables to an empty dict. This would avoid potential syntactic errors while initialising parameters.

**Review by RolandMacDoland — APPROVED** [review]
Thanks @chMoussa LGTM. If you agree with my comment, can you raise an issue about it ?

File: qadence/types.py (hunk 11 above)
Comment (chMoussa) [reply qadence/types.py:433]: Sure, I raised #739
